### ISSUE TITLE
Saving 16B models leads to 137M checkponts?

### ISSUE THREAD

**Issue by ChawDoe** [OpenLMLab/MOSS#156]
Is it something wrong?

**Comment by txsun1997** [comment]
@ChawDoe  具体保存模型的代码是什么

**Comment by ChawDoe** [comment]
![image](https://user-images.githubusercontent.com/25716030/234475398-ec0407ef-e807-4d04-90af-e9dc2638bdd2.png)
The code of saving is the same as your fine-tuning scripts.

**Comment by ChawDoe** [comment]
> ![image](https://user-images.githubusercontent.com/25716030/234475398-ec0407ef-e807-4d04-90af-e9dc2638bdd2.png) The code of saving is the same as your fine-tuning scripts.

@txsun1997

**Comment by txsun1997** [comment]
@xyltt

**Comment by ChawDoe** [comment]
And could you provide us with your training curve? It seems that the model converges very fast on my training set. The accuracy of next token prediction is up to 90% just using less than 1k steps. @txsun1997

**Comment by ChawDoe** [comment]
Did you use the data of traditional NLP tasks as pre-trained data? @txsun1997

**Comment by txsun1997** [comment]
@ChawDoe  `moss-moon-003-sft` was trained on the mixture of FLANv2 (for each task we sampled 64 instances) and `moss-003-sft-data`

**Comment by xyltt** [comment]
> > ![image](https://user-images.githubusercontent.com/25716030/234475398-ec0407ef-e807-4d04-90af-e9dc2638bdd2.png) The code of saving is the same as your fine-tuning scripts.
> 
> @txsun1997

Each "*model_states.pt" file is saved by the corresponding gpu process. You need to use "zero_to_fp32.py" file to merge the model weights into one.

**Comment by ChawDoe** [comment]
> > > ![image](https://user-images.githubusercontent.com/25716030/234475398-ec0407ef-e807-4d04-90af-e9dc2638bdd2.png) The code of saving is the same as your fine-tuning scripts.
> > 
> > 
> > @txsun1997
> 
> Each "*model_states.pt" file is saved by the corresponding gpu process. You need to use "zero_to_fp32.py" file to merge the model weights into one.

@xyltt Yeah, I noticed this script and I tried to run it. But it seems that it take too long minutes to merge checkpoints. Do you have a approximated execution time of it?

**Comment by xyltt** [comment]
> > ![image](https://user-images.githubusercontent.com/25716030/234475398-ec0407ef-e807-4d04-90af-e9dc2638bdd2.png) The code of saving is the same as your fine-tuning scripts.
> 
> @txsun1997



> > > > ![image](https://user-images.githubusercontent.com/25716030/234475398-ec0407ef-e807-4d04-90af-e9dc2638bdd2.png) The code of saving is the same as your fine-tuning scripts.
> > > 
> > > 
> > > @txsun1997
> > 
> > 
> > Each "*model_states.pt" file is saved by the corresponding gpu process. You need to use "zero_to_fp32.py" file to merge the model weights into one.
> 
> @xyltt Yeah, I noticed this script and I tried to run it. But it seems that it take too long minutes to merge checkpoints. Do you have a approximated execution time of it?

Most of the time should be spent loading the model and optimizer states. You can change the below code ```model.save_checkpoint(args.output_dir, global_step)``` into:
```
unwrapped_model = accelerator.unwrap_model(model)

unwrapped_model.save_pretrained(
    args.output_dir,
    is_main_process=accelerator.is_main_process,
    save_function=accelerator.save,
    state_dict=accelerator.get_state_dict(model),
)
```
The above code will gather the model weights from all processes into one and save the model completely. But this option requires consolidation of the weights on one GPU it can be slow and memory-demanding, so only use this feature when needed. Please check the details of the [tutorial](https://huggingface.co/docs/accelerate/usage_guides/deepspeed#saving-and-loading).
